File: app/src/main/java/com/goals/ted/goals/Dates.java
package com.goals.ted.goals;

import android.util.Log;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.List;

/**
 * Created by ted on 7/20/16.
 */
public class Dates extends GoalDates{
    public static long oneDayMilli = 86400000;
    private static final int GOAL_COMPLETE = 0;

    public int convert24To12Format(int hour){
        int formattedHour = hour;
        formattedHour -= 12;
        if(formattedHour < 0){
            formattedHour *= -1;
        }
        return formattedHour;
    }

    public long startToDueDate(){
        long startTime = getStartDate().getTimeInMillis();
        long dueDate = getEndDate().getTimeInMillis();
        if(dueDate > startTime){
            long result = (dueDate - startTime)/oneDayMilli;
            return result;
        }else{
            return 0;
        }
    }
    private long startToDueDateMIlli(){
        long startTime = getStartDate().getTimeInMillis();
        long dueDate = getEndDate().getTimeInMillis();
        return dueDate - startTime;
    }
    private long milliToDueDate(){
        long currentTime = System.currentTimeMillis();
        long dueDate = getEndDate().getTimeInMillis();
        System.out.println("milliToDueDate" + (dueDate - currentTime));
        return dueDate - currentTime;

    }
    public long daysTillDueDate(){
        long currentTime = System.currentTimeMillis();
        long dueDate = getEndDate().getTimeInMillis();
        if(dueDate > currentTime){
            long unmeasuredResults = dueDate - currentTime;
            long results = (dueDate - currentTime)/oneDayMilli;
            if(unmeasuredResults < oneDayMilli && unmeasuredResults > 0){
                return 1;
            }
            return results;
        }else{
            return 0;
        }
    }
    public int percentageComplete(){
        long daysTillDueDate = milliToDueDate();
        long startToDueDate = startToDueDateMIlli();
        int percentCompleted = 0;

        try{
            if(isGoalCompleted(milliToDueDate())){
            percentCompleted = 100;
            }else{
                percentCompleted = calculatePercentage(startToDueDate, daysTillDueDate);
            }
        }catch (ArithmeticException e){
            percentCompleted = 100;
        }

        return percentCompleted;
    }
    private boolean isGoalCompleted(long daysTillDueDate){
        boolean isCompleted;
        if(daysTillDueDate == GOAL_COMPLETE || daysTillDueDate < 0){
            isCompleted = true;
        }else{
            isCompleted = false;
        }
        return isCompleted;
    }
    private int calculatePercentage(long startToDueDate, long daysTillDueDate){
        int percentage = 0;
        int daysCompleted = (int) (startToDueDate - daysTillDueDate);
        percentage = (int) (daysCompleted*100/startToDueDate);
        return percentage;
    }


}

File: app/src/main/java/com/goals/ted/goals/Dates.kt
package com.goals.ted.goals


/**
 * Created by ted on 7/20/16.
 */
open class Dates : GoalDates() {

    fun convert24To12Format(hour: Int): Int {
        var formattedHour = hour
        formattedHour -= 12
        if (formattedHour < 0) {
            formattedHour *= -1
        }
        return formattedHour
    }

    fun startToDueDate(): Long {
        val startTime = startDate!!.timeInMillis
        val dueDate = endDate!!.timeInMillis
        if (dueDate > startTime) {
            val result = (dueDate - startTime) / oneDayMilli
            return result
        } else {
            return 0
        }
    }

    private fun startToDueDateMIlli(): Long {
        val startTime = startDate!!.timeInMillis
        val dueDate = endDate!!.timeInMillis
        return dueDate - startTime
    }

    private fun milliToDueDate(): Long {
        val currentTime = System.currentTimeMillis()
        val dueDate = endDate!!.timeInMillis
        println("milliToDueDate" + (dueDate - currentTime))
        return dueDate - currentTime

    }

    fun daysTillDueDate(): Long {
        val currentTime = System.currentTimeMillis()
        val dueDate = endDate!!.timeInMillis
        if (dueDate > currentTime) {
            val unmeasuredResults = dueDate - currentTime
            val results = (dueDate - currentTime) / oneDayMilli
            if (unmeasuredResults < oneDayMilli && unmeasuredResults > 0) {
                return 1
            }
            return results
        } else {
            return 0
        }
    }

    fun percentageComplete(): Int {
        val daysTillDueDate = milliToDueDate()
        val startToDueDate = startToDueDateMIlli()
        var percentCompleted = 0

        try {
            if (isGoalCompleted(milliToDueDate())) {
                percentCompleted = 100
            } else {
                percentCompleted = calculatePercentage(startToDueDate, daysTillDueDate)
            }
        } catch (e: ArithmeticException) {
            percentCompleted = 100
        }

        return percentCompleted
    }

    private fun isGoalCompleted(daysTillDueDate: Long): Boolean {
        val isCompleted: Boolean = daysTillDueDate == GOAL_COMPLETE.toLong()
                || daysTillDueDate < 0
        return isCompleted
    }

    private fun calculatePercentage(startToDueDate: Long, daysTillDueDate: Long): Int {
        var percentage = 0
        val daysCompleted = (startToDueDate - daysTillDueDate).toInt()
        percentage = (daysCompleted * 100 / startToDueDate).toInt()
        return percentage
    }

    companion object {
        var oneDayMilli: Long = 86400000
        private val GOAL_COMPLETE = 0
    }


}
